src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs

```diff
@@ -328,6 +334,49 @@ protected ulong ClearRegister
             set {  *(ulong*)(_registerViewPointer->GPCLR) = value; }
         }
 
+        /// <summary>
+        /// Returns the pperipheral base address of the raspberry pi based on the ranges set within the device tree.
+        /// The return value is 32bit even on 64 bit operating systems (debian / ubuntu tested) but may change in the future
+        /// 
+        /// Based on file at https://github.com/RPi-Distro/raspi-gpio/blob/master/raspi-gpio.c
+        /// </summary>
+        /// <returns>This reurns the peripheral base address as a 32 bit address.</returns>
+        private uint GetPeripheralBaseAddress()
+        {
+            uint retval = 0xFFFFFFF;
+            byte[] buffer;
+
+            uint firstRange;
+
+            using (BinaryReader rdr = new BinaryReader(File.Open(DeviceTreeRanges, FileMode.Open, FileAccess.Read)))
+            {
+                // get the first bigendian range... this seems to allways be 0x7E000000
+                buffer = rdr.ReadBytes(4);
+                firstRange = ((uint)buffer[0] << 24) | (uint)(buffer[1] << 16) | (uint)(buffer[2] << 8) | buffer[3];
+
+                // read the next range into a buffer....
+                buffer = rdr.ReadBytes(4);
+
+                // if the next range is 0 then assume that the base address is in the next 32 bits so read again.
+                if ((buffer[0] | buffer[1] | buffer[2] | buffer[3]) == 0)
+                {
+                    buffer = rdr.ReadBytes(4);
+                }
+
+                // convert the read data into big endian
+                retval = (uint)(buffer[0] << 24) | (uint)(buffer[1] << 16) | (uint)(buffer[2] << 8) | buffer[3];
+
+                // if the values don't look correct then assume an error and return 0
+                if (firstRange != 0x7E000000 || !(retval == 0x20000000 || retval == 0x3F000000 || retval == 0xFE000000))
+                {
+                    retval = 0xFFFFFFFF;
+                }
+
+            }
+
+            return (retval);
```

Comment:
NIT: remove parenthesis.

Reply:
Well I am checking that the value being returned is consistent with a Pi1,2,3,4 and if not then return a value indicating an error and throw an exception later. It is a double sanity check to ensure that we do not accidentally write to non io space. Tested on Pi 2,3,4 Raspbian, Debian and Ubuntu.

Reply:
Changed.... Will be in the next PR Update